halo2-base/src/utils/halo2.rs

```diff
@@ -113,3 +115,9 @@ pub fn constrain_virtual_equals_external<F: Field + Ord>(
     let acell = copy_manager.assigned_advices.get(&ctx_cell).expect("cell not assigned");
     region.constrain_equal(*acell, external_cell);
 }
+
+/// Circuit with default values.
+pub trait CircuitWithDefault<F: Field>: Circuit<F> {
+    /// Return Circuit with default inputs for the given params.
+    fn default(witness_gen_only: bool, params: Self::Params) -> Self;
```

Comment:
does it even need `witness_gen_only`? do we ever want to use the default circuit with `witness_gen_only = true` (that will require a circuit where constraints are actually satisfied).

Also I was wondering if we want the `params` here to be the same as `Self::Params`. Or if we want to allow a more minimal type like just `k`. (I guess you can always just give dummy and auto-tune.)

Reply:
We will need `witness_gen_only = true` for proof generation benchmarking and `witness_gen_only = false` for proving key generation.

`k` cannot specify which exact circuit we want. e.g. we need output capacity for a component circuit. `Self::params` doesn't need to have tuned `BaseCircuitParams`. Instead it's more like a prompt `BaseCircuitParams` and we can call `calculate_params` inside `default`.